src/ssl.c

```diff
@@ -31622,40 +31550,41 @@ WC_RNG* WOLFSSL_RSA_GetRNG(WOLFSSL_RSA *rsa, WC_RNG **tmpRNG, int *initTmpRng)
  */
 static int wolfSSL_RAND_Init(void)
 {
-    if (wc_LockMutex(&globalRNGMutex) != 0) {
-        WOLFSSL_MSG("Bad Lock Mutex rng");
-        return 0;
-    }
-    if (initGlobalRNG == 0) {
-        if (wc_InitRng(&globalRNG) < 0) {
-            WOLFSSL_MSG("wolfSSL Init Global RNG failed");
-            wc_UnLockMutex(&globalRNGMutex);
-            return 0;
+    int ret = WOLFSSL_FAILURE;
+#ifdef HAVE_GLOBAL_RNG
+    if (wc_LockMutex(&globalRNGMutex) == 0) {
+        if (initGlobalRNG == 0) {
+            ret = wc_InitRng(&globalRNG);
+            if (ret == 0) {
+                initGlobalRNG = 1;
+                ret = WOLFSSL_SUCCESS;
+            }
         }
-        initGlobalRNG = 1;
+        wc_UnLockMutex(&globalRNGMutex);
     }
-
-    wc_UnLockMutex(&globalRNGMutex);
-    return WOLFSSL_SUCCESS;
+#endif
+    return ret;
 }
 
 
 /* WOLFSSL_SUCCESS on ok */
-int wolfSSL_RAND_seed(const void* seed, int len)
+void wolfSSL_RAND_seed(const void* seed, int len)
 {
-    RandSeedMethod* randFunc = NULL;
-
-    WOLFSSL_MSG("wolfSSL_RAND_seed");
-
-    if (GetRandFunction(RAND_SEED, (void**)&randFunc)
-        != WOLFSSL_SUCCESS) {
-        return WOLFSSL_FAILURE;
-    }
-    else if (randFunc != NULL) {
-        return randFunc(seed, len);
+#ifndef WOLFSSL_NO_OPENSSL_RAND_CB
+    if (wolfSSL_RAND_InitMutex() == 0 && wc_LockMutex(&gRandMethodMutex) == 0) {
+        if (gRandMethods && gRandMethods->seed) {
+            /* seed callback has return code, but function RAND_seed does not */
+            (void)gRandMethods->seed(seed, len);
```

Comment:
If we call this function, we should return early instead of calling wolfSSL_RAND_Init.  These functions are supposed to completely override the RAND functions if they are set.

Reply:
Yes but we do not enforce setting all of them and we need to make sure the global rng is initialized. Feel free to improve.